src/goto-programs/goto_program_template.h

```diff
@@ -80,60 +79,129 @@ class goto_program_templatet
     return *this;
   }
 
-  /*! \brief Container for an instruction of the goto-program
-  */
+  /// This class represents a node in the control flow graph of the function.
+  /// The successor nodes are determined by:
+  ///
+  /// - The position of this node in the field
+  ///   `goto_program_templatet::instructions` of the external class. The
+  ///   instruction stored after this instruction is implicitly a successor of
+  ///   this instruction (unless this instruction unconditionally jumps
+  ///   elsewhere).
+  /// - The field `targets` in this class, storing a list of pointers
+  ///   (usually only 1) to other instructions in the CFG.
+  ///
+  /// Three other fields are key:
+  ///
+  /// - type:  an enum value describing the action performed by this instruction
+  /// - guard: an (arbitrarily complex) expression (usually an \ref exprt)
+  /// - code:  a code statement (usually a \ref codet)
+  ///
+  /// The meaning of an instruction node depends on the `type` field. Different
+  /// kinds of instructions make use of the fields `guard` and `code` for
+  /// different purposes.  We list below, using a mixture of pseudo code and
+  /// plain English, the meaning of different kinds of instructions.
+  /// We use `guard`, `code`, and `targets` to mean the value of the
+  /// respective fields in this class:
+  ///
+  /// - GOTO:
+  ///     if `guard` then goto `targets`
+  /// - RETURN:
+  ///     `code` (an instance of code_returnt)
+  /// - DECL:
+  ///     `code` (an instance of code_declt)
+  /// - FUNCTION_CALL:
+  ///     `code` (an instance of code_function_callt)
+  /// - ASSIGN:
+  ///     `code` (an instance of code_assignt)
+  /// - OTHER:
+  ///     `code` (an instance of codet of kind ID_fence, ID_printf,
+  ///     ID_array_copy, ID_array_set, ID_input, ID_output, ...)
+  /// - ASSUME:
+  ///     assume(`guard`)
+  /// - ASSERT:
+  ///     assert(`guard`)
+  /// - SKIP, LOCATION:
+  ///     skip()
+  /// - ATOMIC_BEGIN, ATOMIC_END:
+  ///     There seems to be only 1 mutex in CPROVER, and these instructions
+  ///     acquire or release such mutex.
```

Comment:
Not quite, the guarantee is even stronger: no other instruction will be executing while the code between `ATOMIC_BEGIN`, `ATOMIC_END` is executing.